store/tikv/batch_coprocessor.go

```diff
@@ -0,0 +1,415 @@
+		return rs.respSize
+	}
+
+	// ignore rs.err
+	rs.respSize += int64(cap(rs.startKey))
+	if rs.detail != nil {
+		rs.respSize += int64(sizeofExecDetails)
+		if rs.detail.CommitDetail != nil {
+			rs.respSize += int64(sizeofCommitDetails)
+		}
+	}
+	if rs.pbResp != nil {
+		// Using a approximate size since it's hard to get a accurate value.
+		rs.respSize += int64(rs.pbResp.Size())
+	}
+	return rs.respSize
+}
+
+func (rs *batchCopResponse) RespTime() time.Duration {
+	return rs.respTime
+}
+
+type copTaskAndRPCContext struct {
+	task *copTask
+	ctx  *RPCContext
+}
+
+func buildBatchCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, req *kv.Request) ([]*batchCopTask, error) {
+	if req.StoreType != kv.TiFlash {
+		return nil, errors.New("store type must be tiflash")
+	}
+
+	start := time.Now()
+	const cmdType = tikvrpc.CmdBatchCop
+	rangesLen := ranges.len()
+	for {
+		var tasks []*copTask
+		appendTask := func(regionWithRangeInfo *KeyLocation, ranges *copRanges) {
+			tasks = append(tasks, &copTask{
+				region:    regionWithRangeInfo.Region,
+				ranges:    ranges,
+				cmdType:   cmdType,
+				storeType: req.StoreType,
+			})
+		}
+
+		err := splitRanges(bo, cache, ranges, appendTask)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		var batchTasks []*batchCopTask
+
+		storeTaskMap := make(map[string]*batchCopTask)
+		needRetry := false
+		for _, task := range tasks {
+			rpcCtx, err := cache.GetTiFlashRPCContext(bo, task.region)
+			if err != nil {
+				return nil, err
+			}
+			// If the region is not found in cache, it must be out
+			// of date and already be cleaned up. We should retry and generate new tasks.
+			if rpcCtx == nil {
+				needRetry = true
+			}
+			if batchCop, ok := storeTaskMap[rpcCtx.Addr]; ok {
+				batchCop.copTasks = append(batchCop.copTasks, copTaskAndRPCContext{task: task, ctx: rpcCtx})
+				batchCop.regionTaskMap[task.region.id] = task
+			} else {
+				batchTask := &batchCopTask{
+					storeAddr:     rpcCtx.Addr,
+					cmdType:       cmdType,
+					regionTaskMap: make(map[uint64]*copTask),
+					copTasks:      []copTaskAndRPCContext{{task, rpcCtx}},
+				}
+				batchTask.regionTaskMap[task.region.id] = task
+				storeTaskMap[rpcCtx.Addr] = batchTask
+			}
+		}
+		if needRetry {
+			continue
+		}
+		for _, task := range storeTaskMap {
+			batchTasks = append(batchTasks, task)
+		}
+
+		if elapsed := time.Since(start); elapsed > time.Millisecond*500 {
+			logutil.BgLogger().Warn("buildBatchCopTasks takes too much time",
+				zap.Duration("elapsed", elapsed),
+				zap.Int("range len", rangesLen),
+				zap.Int("task len", len(batchTasks)))
+		}
+		tikvTxnRegionsNumHistogramWithBatchCoprocessor.Observe(float64(len(batchTasks)))
+		return batchTasks, nil
+	}
+}
+
+func (c *CopClient) sendBatch(ctx context.Context, req *kv.Request, vars *kv.Variables) kv.Response {
+	if req.KeepOrder || req.Desc {
+		return copErrorResponse{errors.New("batch coprocessor cannot prove keep order or desc property")}
```

Comment:
Since TiFlash can keep order for handle scan. When we enable the batch cop request, there is none guarantee to keep order for the handle column. So we should ban the order properties' satisfaction in the planner.

Reply:
It's not easy to consider allow_batch in planner. For **safe** and convient, I will check KeepOrder in `executor/builder.go`, If it is true, batch_cop is forced to set flase.

This part of code is remained as proctection.

Reply:
OK, that's also another solution.